### DISCUSSION TITLE
How to create "static" files through plugin?

### DISCUSSION THREAD

**Discussion by rohit-gohri** [facebook/docusaurus#6907] [Q&A] [answered]
I'm currently adding support for multi-file OpenAPI files (https://github.com/rohit-gohri/redocusaurus/issues/23) to my plugin https://github.com/rohit-gohri/redocusaurus. I've got the bundling part done, but to provide a download link for the bundled spec I need to programmatically create a static file that would be included in the build.

AFAIK the `addRoute` option is for creating pages with react and not for static assets. I could copy the file to the static folder but then users of my plugin will have to gitignore that folder.

**Comment by Josh-Cena** [comment, marked as answer]
Just call `fs.writeFile(path.join(context.siteDir, context.siteConfig.staticDirectories[0] ?? 'static'), data)`? 🤔

**Reply by rohit-gohri** [reply]
This would mean users of my plugin would have to manually put the files created in static directory in gitignore, since they are dynamically created.

**Reply by Josh-Cena** [reply]
What's your goal here? To have a static file that's copied to the build output, yet is not tracked by git (i.e. in the default generated folder)?

**Reply by rohit-gohri** [reply]
Yes, exactly. But this static file is programmatically created by the plugin. User is not aware of this file

**Reply by Josh-Cena** [reply]
I think you can use the `postBuild` lifecycle and write directly to the `build` folder? Note that during development, you shouldn't assume the existence of `build`, since all assets are virtually served by webpack dev server. In this case, you either have to use `static`, or you use `configureWebpack().devServer` to proxy certain routes to a virtual HTML file.

**Reply by rohit-gohri** [reply]
> I think you can use the postBuild lifecycle and write directly to the build folder?

Cool, I'll try that. Thanks

> Note that during development, you shouldn't assume the existence of build.

I will skip it in dev, it's not a critical thing

**Reply by rohit-gohri** [reply]
`postBuild` is exactly what I needed. Thanks!

**Reply by slorber** [reply]
The sitemaps plugin is a good example ;)
